hw 1, task 4: make function that returns list filled with powers of 2

furiousteabag opened the pull request.

furiousteabag committed:
Create vozvrspisok.fs

yurii-litvinov commented on vozvrspisok.fs:
```diff
@@ -0,0 +1,25 @@
+// Exponentiation.
```
Тут тоже беда с названием файла

yurii-litvinov commented on vozvrspisok.fs:
```diff
@@ -0,0 +1,25 @@
+    loop list []
+
+// Main function.
+let func n m =
```
func не годится как имя :)

yurii-litvinov commented on vozvrspisok.fs:
```diff
@@ -0,0 +1,25 @@
+
+// Main function.
+let func n m =
+    let rec loop n m (acc:list<int>)=
```
Ну и тут, n и m ничего не говорят о назначении этих штук

yurii-litvinov commented on vozvrspisok.fs:
```diff
@@ -0,0 +1,25 @@
+    let rec loop n m (acc:list<int>)=
+        match acc.Length with
+        | i when i = m + 1 -> rev acc
+        | _ -> loop n m (pow 2 (acc.Length + n)::acc)
```
Не, это работает за квадратичное время (потому что каждый раз заново возводит в степень), можно за линейное.

furiousteabag committed:
Changed filename + minor fixes

furiousteabag committed:
Made algorythm work for linear time

Now we are saving current power and each iteration we just multiplying it by 2.

yurii-litvinov commented on fillListWithPowersOfTwo.fs:
```diff
@@ -0,0 +1,26 @@
+    loop list []
+
+// Returns list filled with powers of 2.
+let fillListWithPowersOfTwo powerToStart numberOfElememtsMinusOne =
```
numberOfEleme**m**tsMinusOne, опечатка

yurii-litvinov commented on fillListWithPowersOfTwo.fs:
```diff
@@ -0,0 +1,26 @@
+        match list.Length with
+        | i when i = numberOfElememtsMinusOne + 1 -> rev list
+        | 0 -> loop powerToStart numberOfElememtsMinusOne (pow 2 (list.Length + powerToStart)::list) (pow 2 (list.Length + powerToStart))
+        | _ -> loop powerToStart numberOfElememtsMinusOne (2*acc::list) (2*acc)
```
Надо вокруг `*` пробелы

yurii-litvinov commented on fillListWithPowersOfTwo.fs:
```diff
@@ -0,0 +1,26 @@
+    let rec loop powerToStart numberOfElememtsMinusOne (list:list<int>) acc=
+        match list.Length with
+        | i when i = numberOfElememtsMinusOne + 1 -> rev list
+        | 0 -> loop powerToStart numberOfElememtsMinusOne (pow 2 (list.Length + powerToStart)::list) (pow 2 (list.Length + powerToStart))
```
`list.Length + powerToStart` --- хм, но list.Length же равно 0, судя по условию match-а. И какой смысл `pow 2 (list.Length + powerToStart)` считать дважды? Можно
```fsharp
| 0 -> let p = pow 2 (list.Length + powerToStart) 
       loop powerToStart numberOfElememtsMinusOne (p::list) p
```

furiousteabag replied:
Понял, не пришло в голову что после стрелочки можно несколько выражений ставить.

yurii-litvinov reviewed (changes requested):
Надо ещё немного доисправлять

furiousteabag committed:
Update fillListWithPowersOfTwo.fs

yurii-litvinov reviewed (approved):
Коментарий к коммиту не очень. В остальном ок, зачтена.

furiousteabag committed:
Added tests and appveyor.yml
